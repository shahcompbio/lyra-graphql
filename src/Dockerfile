FROM node:8

RUN npm install pm2@latest -g

RUN git clone https://github.com/shahcompbio/lyra-graphql.git && \
  cd lyra-graphql && \
  yarn install && \
  yarn build

EXPOSE 4000

CMD node lyra-graphql/lib/index.js
