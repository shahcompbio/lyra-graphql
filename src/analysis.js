const { gql } = require("apollo-server");

import client from "./api/elasticsearch.js";

export const schema = gql`
  extend type Query {
    dashboards: [Dashboard!]!
    analysis(analysis: String!, dashboard: String!): Analysis
  }

  type Dashboard {
    id: String!
    analyses: [Analysis!]!
  }

  type Analysis {
    id: String!
    title: String!
    description: String!
    segsIndex: String!
    treeIndex: String!
  }
`;

const allAnalysisQuery = () => ({
  size: 50000
});

export const resolvers = {
  Query: {
    async dashboards() {
      const results = await client.search({
        index: "analysis",
        body: {
          size: 50000,
          aggs: {
            dashboards: {
              terms: {
                field: "dashboard",
                size: 50000
              }
            }
          }
        }
      });
      return results.aggregations.dashboards.buckets.map(bucket => ({
        id: bucket.key,
        analyses: results.hits.hits.map(hit => hit["_source"])
      }));
    },
    async analysis(_, { analysis, dashboard }) {
      const results = await client.search({
        index: "analysis",
        body: {
          query: {
            bool: {
              must: [
                {
                  term: {
                    dashboard: {
                      value: dashboard
                    }
                  }
                },
                {
                  term: {
                    analysis_id: {
                      value: analysis
                    }
                  }
                }
              ]
            }
          }
        }
      });

      return results.hits.hits[0]["_source"];
    }
  },
  Dashboard: {
    id: root => root.id,
    analyses: root => {
      const dashboard = root.id;
      return root.analyses.filter(analysis => analysis.dashboard === dashboard);
    }
  },

  Analysis: {
    id: root => root.analysis_id,
    title: root => root.title,
    description: root => root.description,
    segsIndex: root => root.segs_index,
    treeIndex: root => root.tree_index
  }
};
