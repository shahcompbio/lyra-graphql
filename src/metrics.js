const { gql } = require("apollo-server");

import client from "./api/elasticsearch.js";

export const schema = gql`
  extend type Query {
    hasPloidy(analysis: String!): Boolean
  }
`;

export const resolvers = {
  Query: {
    async hasPloidy(_, { analysis }) {
      const results = await client.indices.exists({
        index: `ce00_${analysis.toLowerCase()}_qc`
      });

      return results;
    }
  }
};
