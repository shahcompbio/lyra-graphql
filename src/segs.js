const { gql } = require("apollo-server");

import client from "./api/elasticsearch.js";

export const schema = gql`
  extend type Query {
    chromosomes(analysis: String!): [Chromosome]
    segs(analysis: String!, indices: [Int!]!): [SegRow]
    cloneSegs(analysis: String!, range: [Int!]!): [Seg]
  }

  type Chromosome {
    id: String!
    start: Int!
    end: Int!
  }

  type SegRow {
    id: String!
    index: Int!
    ploidy: Int!
    segs: [Seg!]!
  }

  type Seg {
    chromosome: String!
    start: Int!
    end: Int!
    state: Int!
    integerMedian: Float!
  }
`;

async function getIDsForIndices(analysis, indices) {
  const results = await client.search({
    index: `ce00_${analysis.toLowerCase()}_tree`,
    body: {
      size: 50000,
      sort: [
        {
          heatmap_order: {
            order: "asc"
          }
        }
      ],
      query: {
        bool: {
          must: [
            {
              terms: {
                heatmap_order: indices
              }
            }
          ]
        }
      }
    }
  });

  return results;
}

async function getIDsForRange(analysis, range) {
  const results = await client.search({
    index: `ce00_${analysis.toLowerCase()}_tree`,
    body: {
      size: 50000,
      query: {
        bool: {
          must: [
            {
              range: {
                heatmap_order: {
                  gte: range[0],
                  lte: range[1]
                }
              }
            }
          ]
        }
      }
    }
  });
  return results;
}

const getBinQuery = ids => {
  const rangeQuery = Array.apply(null, { length: 500 }).map(
    (_, i) => ({
      from: i === 0 ? 0 : 500000 * i + 1,
      to: (i + 1) * 500000
    }),
    []
  );

  return {
    size: 0,
    query: {
      bool: {
        must: [
          {
            terms: {
              cell_id: ids
            }
          }
        ]
      }
    },
    aggs: {
      chromosomes: {
        terms: {
          field: "chrom_number",
          size: 50
        },
        aggs: {
          bins: {
            histogram: {
              field: "start",
              interval: 500000
            },
            aggs: {
              state: {
                terms: {
                  field: "state",
                  size: 1
                }
              }
            }
          }
        }
      }
    }
  };
};

const getSegsForChromosome = (id, bins) => {
  const convertBinsToSegs = (currSeg, allSegs, bins) => {
    if (bins.length === 0) {
      return [...allSegs, currSeg];
    } else {
      const [firstBin, ...restBin] = bins;

      if (hasSameState(currSeg, firstBin)) {
        return convertBinsToSegs(
          mergeSegs(currSeg, firstBin),
          allSegs,
          restBin
        );
      } else {
        return convertBinsToSegs(
          convertToSeg(firstBin),
          [...allSegs, currSeg],
          restBin
        );
      }
    }
  };

  const getBinState = bin => bin.state.buckets[0].key;

  const hasSameState = (seg, bin) => seg.state === getBinState(bin);

  const mergeSegs = (seg, bin) => ({
    ...seg,
    end: bin.key + 500000
  });

  const convertToSeg = bin => ({
    chrom_number: id,
    start: bin.key === 0 ? bin.key : bin.key + 1,
    end: bin.key + 500000,
    state: getBinState(bin)
  });
  const [firstBin, ...restBin] = bins;
  return convertBinsToSegs(convertToSeg(firstBin), [], restBin);
};
export const resolvers = {
  Query: {
    async chromosomes(_, { analysis }) {
      const results = await client.search({
        index: `ce00_${analysis.toLowerCase()}_segs`,
        body: {
          size: 0,
          aggs: {
            chrom_ranges: {
              terms: {
                field: "chrom_number",
                size: 50000,
                order: {
                  _key: "asc"
                }
              },
              aggs: {
                XMax: {
                  max: {
                    field: "end"
                  }
                },
                XMin: {
                  min: {
                    field: "start"
                  }
                }
              }
            }
          }
        }
      });

      return results.aggregations.chrom_ranges.buckets;
    },

    async segs(_, { analysis, indices }) {
      const results = await getIDsForIndices(analysis, indices);

      return results.hits.hits;
    },

    async cloneSegs(_, { analysis, range }) {
      const idResults = await getIDsForRange(analysis, range);
      const ids = idResults.hits.hits.map(record => record["_source"].cell_id);

      const cloneBins = await client.search({
        index: `cl00_${analysis.toLowerCase()}_bins`,
        body: getBinQuery(ids)
      });

      const results = cloneBins.aggregations.chromosomes.buckets.reduce(
        (results, chromosome) => [
          ...results,
          ...getSegsForChromosome(chromosome.key, chromosome.bins.buckets)
        ],
        []
      );

      const roundedResults = results.map(result => ({
        ...result,
        state: Math.round(result.state)
      }));
      return results;
    }
  },

  Chromosome: {
    id: root => root.key,
    start: root => root.XMin.value,
    end: root => root.XMax.value
  },

  SegRow: {
    id: root => root["_source"].cell_id,
    index: root => root["_source"].heatmap_order,
    ploidy: async root => {
      const ploidyIndex = root["_index"].replace("_tree", "_qc");

      const isIndexExist = await client.indices.exists({
        index: ploidyIndex
      });

      if (isIndexExist) {
        const results = await client.search({
          index: ploidyIndex,
          body: {
            size: 50000,
            query: {
              bool: {
                filter: [{ term: { cell_id: root["_source"].cell_id } }]
              }
            }
          }
        });

        const hit = results.hits.hits;
        return hit.length === 0 ? -1 : hit[0]["_source"]["state_mode"];
      }
      return -1;
    },
    segs: async root => {
      const results = await client.search({
        index: root["_index"].replace("_tree", "_segs"),
        body: {
          size: 50000,
          query: {
            bool: {
              filter: [{ term: { cell_id: root["_source"].cell_id } }]
            }
          }
        }
      });

      return results.hits.hits.map(seg => seg["_source"]);
    }
  },

  Seg: {
    chromosome: root => root.chrom_number,
    start: root => root.start,
    end: root => root.end,
    state: root => root.state,
    integerMedian: root => root.integer_median
  }
};
