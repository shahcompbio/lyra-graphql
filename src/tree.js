const { gql } = require("apollo-server");

import client from "./api/elasticsearch.js";

export const schema = gql`
  extend type Query {
    treeRoot(analysis: String!): Node
    treeNode(analysis: String!, id: String, index: Int): Node
    treeNodes(analysis: String!, range: [Int!]!): [Node]
  }

  type Node {
    id: [String]!
    parent: String!
    index: Int!
    maxIndex: Int!
    maxHeight: Int!
    children: [NodeChild!]!
  }

  type NodeChild {
    id: String!
    index: Int!
    maxIndex: Int!
    maxHeight: Int!
  }
`;

export const resolvers = {
  Query: {
    async treeRoot(_, { analysis }) {
      const results = await client.search({
        index: `ce00_${analysis.toLowerCase()}_tree`,
        body: {
          size: 1,
          query: {
            bool: {
              filter: [{ term: { parent: "root" } }]
            }
          }
        }
      });

      return results.hits.hits[0];
    },

    async treeNode(_, { analysis, id, index }) {
      const term = id ? { cell_id: id } : { heatmap_order: index };

      const results = await client.search({
        index: `ce00_${analysis.toLowerCase()}_tree`,
        body: {
          size: 1,
          query: {
            bool: {
              filter: [{ term: term }]
            }
          }
        }
      });
      return results.hits.hits[0];
    },

    async treeNodes(_, { analysis, range }) {
      const [minIndex, maxIndex] = range;

      const results = await client.search({
        index: `ce00_${analysis.toLowerCase()}_tree`,
        body: {
          size: 50000,
          sort: [
            {
              heatmap_order: {
                order: "asc"
              }
            }
          ],
          query: {
            bool: {
              must: [
                {
                  range: {
                    heatmap_order: {
                      gte: minIndex,
                      lte: maxIndex
                    }
                  }
                }
              ]
            }
          }
        }
      });

      return results.hits.hits;
    }
  },
  Node: {
    id: root => root["_source"].cell_id.split(",").map(item => item.trim()),
    parent: root => root["_source"].parent,
    index: root => root["_source"].heatmap_order,
    maxIndex: root => root["_source"].max_index,
    maxHeight: root => root["_source"].max_height,
    children: root => {
      return root["_source"].children.map(async child => {
        const results = await client.search({
          index: root["_index"],
          body: {
            size: 1,
            query: {
              bool: {
                filter: [{ term: { unmerged_id: child } }]
              }
            }
          }
        });
        return results.hits.hits[0]["_source"];
      });
    }
  },

  NodeChild: {
    id: root => root.cell_id,
    index: root => root.heatmap_order,
    maxIndex: root => root.max_index,
    maxHeight: root => root.max_height
  }
};
