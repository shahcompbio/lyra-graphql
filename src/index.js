import "babel-polyfill";
import * as tree from "./tree.js";
import * as segs from "./segs.js";
import * as metrics from "./metrics.js";
import { merge } from "lodash";

import client from "./api/elasticsearch.js";

const { ApolloServer, gql } = require("apollo-server");

export const schema = gql`
  type Query {
    dashboards: [Dashboard!]!
    analysis(analysis: String!, dashboard: String!): Analysis
  }

  type Dashboard {
    id: String!
    analyses: [Analysis!]!
  }

  type Analysis {
    id: String!
    title: String!
    description: String!
    segsIndex: String!
    treeIndex: String!
  }
`;

const allAnalysisQuery = () => ({
  size: 50000
});

export const resolvers = {
  Query: {
    async dashboards() {
      const results = await client.search({
        index: "analysis",
        body: {
          size: 50000,
          aggs: {
            dashboards: {
              terms: {
                field: "dashboard",
                size: 50000
              }
            }
          }
        }
      });
      return results.aggregations.dashboards.buckets.map(bucket => ({
        id: bucket.key,
        analyses: results.hits.hits.map(hit => hit["_source"])
      }));
    },
    async analysis(_, { analysis, dashboard }) {
      const results = await client.search({
        index: "analysis",
        body: {
          query: {
            bool: {
              must: [
                {
                  term: {
                    dashboard: {
                      value: dashboard
                    }
                  }
                },
                {
                  term: {
                    analysis_id: {
                      value: analysis
                    }
                  }
                }
              ]
            }
          }
        }
      });

      return results.hits.hits[0]["_source"];
    }
  },
  Dashboard: {
    id: root => root.id,
    analyses: root => {
      const dashboard = root.id;
      return root.analyses.filter(analysis => analysis.dashboard === dashboard);
    }
  },

  Analysis: {
    id: root => root.analysis_id,
    title: root => root.title,
    description: root => root.description,
    segsIndex: root => root.segs_index,
    treeIndex: root => root.tree_index
  }
};

const server = new ApolloServer({
  typeDefs: [schema, tree.schema, segs.schema, metrics.schema],
  resolvers: merge(
    resolvers,
    tree.resolvers,
    segs.resolvers,
    metrics.resolvers
  ),
  mocks: true
});

const express = require("express");
const { registerServer } = require("apollo-server-express");
const app = express();
registerServer({ server, app, path: "/graphql" });

server.listen().then(({ url }) => console.log(`Server is running on ${url}`));
