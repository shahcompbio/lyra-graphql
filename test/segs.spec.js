import "@babel/polyfill";
import { schema, resolvers } from "../src/segs";
import { startServer, stopServer } from "./test-server";
import { gql } from "apollo-server";

const ANALYSIS_ID = "ABC_123";

let query;
beforeAll(async () => {
  const client = await startServer(schema, resolvers);
  query = client.query;
});

it("has server started", () => {
  expect(query).toBeDefined();
});

it("test chromosomes query", async () => {
  const CHROMOSOMES_QUERY = gql`
    query {
      chromosomes {
        id
        start
        end
      }
    }
  `;

  const res = await query({
    query: CHROMOSOMES_QUERY,
    variables: { analysis: ANALYSIS_ID }
  });
  expect(res).toMatchSnapshot();
});

it("test segs query", async () => {
  const SEGS_QUERY = gql`
    query {
      segs {
        id
        name
        index
        ploidy
        segs {
          chromosome
          start
          end
          state
        }
      }
    }
  `;

  const res = await query({
    query: SEGS_QUERY,
    variables: { analysis: ANALYSIS_ID, indices: [0, 2] }
  });
  expect(res).toMatchSnapshot();
});
