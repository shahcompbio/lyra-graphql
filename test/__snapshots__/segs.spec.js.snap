// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test chromosomes query 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    [ValidationError: Field "chromosomes" argument "analysis" of type "String!" is required but not provided.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;

exports[`test segs query 1`] = `
Object {
  "data": undefined,
  "errors": Array [
    [ValidationError: Field "segs" argument "analysis" of type "String!" is required but not provided.],
    [ValidationError: Field "segs" argument "indices" of type "[Int!]!" is required but not provided.],
    [ValidationError: Field "segs" argument "isNorm" of type "Boolean!" is required but not provided.],
  ],
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
